File: Math challenge/Challanges/Subtraction.m
//
//  Subtraction.m
//  Math challenge
//
//  Created by Lukas Bischof on 09.01.18.
//  Copyright © 2018 Lukas Bischof. All rights reserved.
//

#import "Subtraction.h"

@implementation Subtraction {
    NSInteger _level;
}

+ (instancetype)newWithLevel:(NSInteger)level
{
    return [[Subtraction alloc] initWithLevel:level];
}

- (instancetype)initWithLevel:(NSInteger)level
{
    if ((self = [super init])) {
        _level = level;
    }
    
    return self;
}

- (Challange *)next {
    NSInteger lowerBound = (NSInteger)powf(10.f, (float)_level);
    NSInteger upperBound = (NSInteger)powf(10.f, (float)(_level + 1)) - 1;
    
    NSInteger firstTmp = randr(lowerBound, upperBound);
    NSInteger secondTmp = randr(lowerBound, upperBound);
    
    NSInteger first = MAX(firstTmp, secondTmp);
    NSInteger second = MIN(firstTmp, secondTmp);
    
    return [[Challange alloc] initWithText:SWF(@"%ld - %ld", (long)first, (long)second) andAnswer:first - second];
}

- (NSString *)challangeDescription
{
    return SWF(@"Subtraktion level %ld", (long)_level);
}

@end

File: Math challenge/challenges/Subtraction.swift
//
// Created by Lukas Bischof on 25.10.19.
// Copyright (c) 2019 Lukas Bischof. All rights reserved.
//

import Foundation

class Subtraction: MathChallengeGenerator {
  var challengeDescription: String {
    get { "Subtraktion von \(lowerBound!) bis \(upperBound!)" }
  }

  private let level: Int
  private var lowerBound: Int!
  private var upperBound: Int!

  init(level: Int) {
    self.level = level
    self.lowerBound = bound(level - 1)
    self.upperBound = bound(level) - 1
  }

  func next() -> Challenge {
    let result = randomInt(lowerBound, upperBound)
    let right = randomInt(lowerBound, upperBound)
    let left = result + right

    return Challenge(text: "\(left) - \(right)", andAnswer: result)
  }

  private func bound(_ upper: Int) -> Int {
    return Int(powf(10.0, Float(upper)))
  }
}
